java/debugger.jpda/src/org/netbeans/modules/debugger/jpda/models/ShortenedStrings.java

```diff
@@ -385,19 +349,220 @@ else if (valuesField.type() instanceof ArrayType &&
+                index * 2, 2);
+        Value left = vals.get(0);
+        Value right = vals.get(1);
+        if (!(left instanceof ByteValue && right instanceof ByteValue)){
+            return -1;
+        }
+        return utf16Combine(((ByteValue)left).byteValue(),
+                ((ByteValue)right).value(), isLittleEndian);
+    }
+    /**
+     * (Currently untested) Grab the char at the given index in the array.
+     * Returns -1 on an error
+     * @param sourceArray Backing array reference. May be a byte or char array
+     * @param index
+     * @param encoding
+     * @param isLittleEndian
+     * @return
+     */
+    private static int charAt(List<Value> sourceArray, int index,
+            InternalStringEncoding encoding, boolean isLittleEndian){
+        if (encoding == InternalStringEncoding.CHAR_ARRAY){
+            //that was easy
+            Value v = sourceArray.get(index);
+            if (!(v instanceof CharValue)){
+                return -1;
+            }
+            return ((CharValue)v).charValue();
+        }
+        if (encoding == InternalStringEncoding.BYTE_ARRAY_LATIN1){
+            //that was also easy
+            Value v = sourceArray.get(index);
+            if (!(v instanceof ByteValue)){
+                return -1;
+            }
+            char c = (char)((ByteValue)v).byteValue();
+            //strip off the sign value
+            c &= 0xff;
+            return c;
+        }
+        //uft16 it is
+        Value left = sourceArray.get(index * 2);
+        Value right = sourceArray.get((index * 2) + 1);
+        if (!(left instanceof ByteValue && right instanceof ByteValue)){
+            return -1;
+        }
+        return utf16Combine(((ByteValue)left).byteValue(),
+                ((ByteValue)right).value(), isLittleEndian);
+    }
+    /**
+     * Copy the input to the destination array as if by {@link System#arrayCopy}
+     * @param sourceArray Backing array reference. May be a byte or char array
+     * @param encoding
+     * @param isLittleEndian
+     * @param start
+     * @param length
+     * @param dest
+     */
+    private static void copyToCharArray(ArrayReference sourceArray,
+            int srcPos, char[] dest, int destPos, int length,
+            InternalStringEncoding encoding, boolean isLittleEndian) throws
+            ObjectCollectedExceptionWrapper, VMDisconnectedExceptionWrapper,
+            InternalExceptionWrapper, IOException{
+        //grab applicable values
+        int realStart = srcPos;
+        int realLength = length;
+        if (encoding == InternalStringEncoding.BYTE_ARRAY_UTF16){
+            realStart *= 2;
+            realLength *= 2;
+        }
+        List<Value> values = ArrayReferenceWrapper.getValues(sourceArray,
+                realStart, realLength);
+        //copy them
+        copyToCharArray(values, 0, dest, destPos, length, encoding,
+                isLittleEndian);
+    }
+    /**
```

Comment:
Please also make sure, that there is an empty line between the end of the previous method and the comment.